lab 5 done

Suetolog opened the pull request:
lab 5 done

Suetolog committed:
lab 5 done

lab 5 done

dvorotnev commented on cvlib/src/corner_detector.cpp:
```diff
@@ -107,8 +108,9 @@ bool corner_detector_fast::check_count_same_in_a_row(int* arr, int thresh)
 }
 
-void corner_detector_fast::compute(cv::InputArray, std::vector<cv::KeyPoint>& keypoints, cv::OutputArray descriptors)
+void corner_detector_fast::compute(cv::InputArray image, std::vector<cv::KeyPoint>& keypoints, cv::OutputArray descriptors)
```
Главный вопрос: если я вызову эту функцию 2 раза подряд для одних и тех же особых точек, у меня совпадут дескрипторы?

Suetolog replied:
Намёк понятен. Если эту функцию вызвать 2 раза подряд, то дескрипторы не совпадут, т.к. создание случайных пар происходит при каждом вызове compute. Вынес вектор случайных пар в приватный член класса и ныне произвожу его инициализацию только один раз. Так же была проблема в том, что при использовании размытия по гауссу я размывал именно то изображение, которое передавал в функцию а не его копию (я уже забыл как работает OpenCV и каким специфическим типом является Mat. Вот щас вспомнил, что при передачи Mat в функцию на самом деле не создается его копия а передается указатель на массив). Таким образом при каждом вызове compute я все больше и больше размывал изображение). Сейчас все проблемы пофиксил. Прикладываю гистограмму расстояний двух дескрипторов моих, посчитанных для одних и тех же особых точек функцией compute. Так же на отдельную кнопку вынес сохранение тех же самых дескрипторов в файл и этот файл прикладываю.
![image](https://user-images.githubusercontent.com/61085239/201153063-05ae3b8f-6d9c-4b43-95b6-d8c9e3a7d9d1.png)
[test.txt](https://github.com/Suetolog/cvclasses19/files/9982894/test.txt)

Suetolog replied:
Сделал коммит исправленной версии

Suetolog committed:
fix

fixing bags

dvorotnev commented on cvlib/src/corner_detector.cpp:
```diff
@@ -35,11 +36,11 @@ void corner_detector_fast::detect(cv::InputArray image, CV_OUT std::vector<cv::K
     cv::Mat img_mat = image.getMat();
     // \todo implement FAST with minimal LOCs(lines of code), but keep code readable.
-
+    cv::blur(img_mat, img_mat, cv::Size(5, 5)); //Дополнительно добавлено гауссово сглаживания для устранения шумов
```
Это сглаживание тоже лучше перенести на уровень выше, чтобы не портить входное изображение

Suetolog replied:
Размывать на уровне выше = так же портить изображение. Так же можно забыть размыть изображение перед вызовом функции и получить милион особых точек. Поэтому выбрал не самый отпимальный путь, зато рабочий. Переделал на работу с копией изображения, а не с исходным.

dvorotnev replied:
:+1:

dvorotnev commented on cvlib/src/corner_detector.cpp:
```diff
@@ -125,6 +127,96 @@ void corner_detector_fast::compute(cv::InputArray, std::vector<cv::KeyPoint>& ke
+    image.getMat().copyTo(img);
+    temp_desc_mat.setTo(0);
+
+    cv::GaussianBlur(img, img, cv::Size(5, 5), 2, 2); //Размытие по гаусу
```
И тут тоже, либо надо создавать копию изображения/области локальной точки

Suetolog replied:
Тут же видно, что я создал копию мата и его размываю.

dvorotnev replied:
Да, точно, не увидел

dvorotnev commented:
Когда исправите - можете мёржить, оценку поставил

Suetolog committed:
fix

Теперь размывается не исходное изображение а его копия, и ведется работа теперь с ней.
